csrc/device_lower/validation.cpp

```diff
@@ -515,6 +522,35 @@ class VectorizeValidator : public OptInDispatch {
   }
 };
 
+// Gather info about starting offsets of accesses to a
+// tensor. Currently only slice is considered.
+std::unordered_set<Val*> getTensorOffsets(TensorView* tv) {
```

Comment:
Hmm, I am hesitant about this. First of all, depending on the contiguity and allocation domain of the tensor, the `stride` here is not actually be the stride of the tensor. This is even more likely to be the case when `tv` is not a fusion input, because there might be a permute/broadcast/reduction on the path from the input to the sliced tensor. Second, even if all the root domain of `slice_use` exactly map to `tv` with the same order, if there are multiple slices, why are we only validating the innermost slice? Third, ignore the two problems above and assume that `stride` is indeed the stride, checking on `offsets` is in a non-innermost dimension is just a no-op, because we are already validating the strides of the input tensors somewhere else, and `mulExpr(slice_info.at(i).start, stride)` is always a multiple of the stride, so if the strides already pass the validation, so will the offsets. In summary, I think this is a difficult question, it is as hard as the reason why Christian disabled part of the vectorize validation. I think for the case of slice, a better strategy might be just not validate it.

Reply:
Sorry I should have added more comments to clarify what I implicitly assumed.

The information collected here is only used for validating vectorization of input slicing. So, we could just skip if `tv` is not a fusion input. And if the innermost sliced domain is involved in the vectorization, that domain as well as any other inner domains must be contiguous. Yes, I neglected allocation domains, but as long as the sliced domain is involved in vectorization, it should mean those domains are contiguous, thus the stride calculation should be correct, right?

> Second, even if all the root domain of slice_use exactly map to tv with the same order, if there are multiple slices, why are we only validating the innermost slice?

This part I'm not really confident now. I think what I was thinking about was something like: 

- the vectorization factor is only derived from the innermost sliced domain since slicing implies its outer domain becomes non contiguous, so no outer sliced domains should be involved in vectorization. 
- The total extent of the inner root domains involved in vectorization should be divisible by the vectorization factor. 

So as long as the innermost sliced domain is aligned, outer domains should not matter. I think that was my thinking. I need to check again the original PR if I put all required checks in the scheduler, though.

> checking on offsets is in a non-innermost dimension is just a no-op,

Suppose we have something like:

```
t0: [10, 2]
auto t1 = t0[0:, :] // just a copy
auto t2 = t0[1:, :]
auto t3 = t0[2:, :]
```
Here, `t1` and `t3` should be vectorizable by only 4 but `t2` should be done just with `2`. Doesn't this mean we should also check split offsets?

Reply:
> as long as the sliced domain is involved in vectorization, it should mean those domains are contiguous

Why? For example, if I have:
```python
>>> import torch
>>> t = torch.empty_strided((1024, 1024), (2048,  1)).zero_()
>>> t.shape, t.stride()
(torch.Size([1024, 1024]), (2048, 1))
>>> t[1:2, :]
tensor([[0., 0., 0.,  ..., 0., 0., 0.]])
```
for this case, it is totally fine for slicing, but the tensor is not contiguous.

Reply:
> > as long as the sliced domain is involved in vectorization, it should mean those domains are contiguous
> 
> Why? For example, if I have:
> 
> ```python
> >>> import torch
> >>> t = torch.empty_strided((1024, 1024), (2048,  1)).zero_()
> >>> t.shape, t.stride()
> (torch.Size([1024, 1024]), (2048, 1))
> >>> t[1:2, :]
> tensor([[0., 0., 0.,  ..., 0., 0., 0.]])
> ```
> 
> for this case, it is totally fine for slicing, but the tensor is not contiguous.

Yeah, in this case, it's vectorizable, and there's no need to check about the slice offset. I should have limited this check for cases where vectorization is done with merged domains with sliced domains. 

Started feeling the same as:
> I think for the case of slice, a better strategy might be just not validate it for now.

Reply:
> Here, `t1` and `t3` should be vectorizable by only 4 but `t2` should be done just with `2`. Doesn't this mean we should also check split offsets?

Oh, yeah, I agree. It's not a no-op. But there is another problem:

I agree that `t3` can be vectorized to 4. But if instead I have `auto t4 = t0[2:9, :]`, still, we will just be able to vectorize 2, but we are lacking of such a check right now I think?

Reply:
> > Here, `t1` and `t3` should be vectorizable by only 4 but `t2` should be done just with `2`. Doesn't this mean we should also check split offsets?
> 
> Oh, yeah, I agree. It's not a no-op. But there is another problem:
> 
> I agree that `t3` can be vectorized to 4. But if instead I have `auto t4 = t0[2:9, :]`, still, we will just be able to vectorize 2, but we are lacking of such a check right now I think?

In that case, the offset would be `2 * stride == 20`, so vectorization by 2 should pass. It's not vectorizable by 4, but I'm not sure if the scheduler analysis in the original PR correctly gives 2 instead of 4.

Reply:
> * its outer domain becomes non contiguous

I think slice can indeed make it possible to vectorize more. For example, if I have a tensor `t0` whose shape is (1024, 1024) and whose stride is `(3, 1024)`, without slice, I can not vectorize, because the first stride is not an odd prime number. However, if I do have a `t0[0:1, :]`, then I can vectorize 4 because the first dim becomes broadcast, so who cares about its stride. I don't think we are taking advantage of this to vectorize 4 (I guess?), but as a validation, we should not reject this.

Reply:
> I think slice can indeed make it possible to vectorize more.

Another example is, if I have:
```
t0: [1025]
t1 = t0 // no vectorization
t2 = t0[:1024] // vectorize 4!
```